### PR TITLE
refactor: Update all engines to use Query and Record dataclasses

### PR DESCRIPTION
**Pull request by KShivendu**
This PR does two things:
- Use `Query` and `Record` in all the engines. 
- Introduces [Ruff](https://docs.astral.sh/ruff/) linter (written in Rust :crab: ) in precommit hooks to run type checks and help maintain code quality.

Part of sparse vectors CI benchmark PR. https://github.com/qdrant/vector-db-benchmark/pull/114

### CODE REVIEW

File: engine/clients/milvus/upload.py [hunk 1]
```diff
@@ -41,20 +42,22 @@ def init_client(cls, host, distance, connection_params, upload_params):
...
+                    or DTYPE_DEFAULT[field_schema.dtype]
+                    for record in batch
                 ]
                 for field_schema in cls.collection.schema.fields
                 if field_schema.name not in ["id", "vector"]
             ]
         else:
             field_values = []
+        ids = [record.idx for record in batch]
+        vectors = [record.vector for record in batch]
```
Comment (joein) [inline engine/clients/milvus/upload.py:60]: can be done with one loop

File: engine/clients/weaviate/upload.py [hunk 2]
```diff
@@ -28,14 +29,14 @@ def init_client(cls, host, distance, connection_params, upload_params):
...
-        cls, ids: List[int], vectors: List[list], metadata: List[Optional[dict]]
-    ):
+    def upload_batch(cls, batch: List[Record]):
         objects = []
-        for i in range(len(ids)):
-            id = uuid.UUID(int=ids[i])
-            property = metadata[i] or {}
-            objects.append(DataObject(properties=property, vector=vectors[i], uuid=id))
+        for record in batch:
+            id = uuid.UUID(int=record.id)
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply engine/clients/weaviate/upload.py]: Fixed. Thanks!

File: engine/clients/elasticsearch/upload.py [hunk 3]
```diff
@@ -44,19 +45,12 @@ def init_client(cls, host, distance, connection_params, upload_params):
...
-        cls, ids: List[int], vectors: List[list], metadata: Optional[List[dict]]
-    ):
-        if metadata is None:
-            metadata = [{}] * len(vectors)
+    def upload_batch(cls, batch: List[Record]):
         operations = []
-        for idx, vector, payload in zip(ids, vectors, metadata):
-            vector_id = uuid.UUID(int=idx).hex
+        for record in batch:
+            vector_id = uuid.UUID(int=record.idx).hex
```
Comment (joein) [inline engine/clients/elasticsearch/upload.py:51]: it's `record.id`, not `record.idx`

File: engine/clients/milvus/upload.py [hunk 4]
```diff
@@ -41,20 +42,27 @@ def init_client(cls, host, distance, connection_params, upload_params):
...
                 ]
                 for field_schema in cls.collection.schema.fields
                 if field_schema.name not in ["id", "vector"]
             ]
         else:
             field_values = []
+
+        ids, vectors = [], []
+        for record in batch:
+            ids.append(record.idx)
```
Comment (joein) [inline engine/clients/milvus/upload.py:63]: `.id`

File: engine/clients/milvus/upload.py [hunk 5]
```diff
@@ -41,20 +42,27 @@ def init_client(cls, host, distance, connection_params, upload_params):
         cls.distance = DISTANCE_MAPPING[distance]
 
     @classmethod
-    def upload_batch(
-        cls, ids: List[int], vectors: List[list], metadata: Optional[List[dict]]
-    ):
-        if metadata is not None:
+    def upload_batch(cls, batch: List[Record]):
+        has_metadata = any(record.metadata for record in batch)
+        field_values = []
```
Comment (joein) [inline engine/clients/milvus/upload.py:47]: defined not used

File: engine/clients/opensearch/search.py [hunk 6]
```diff
@@ -46,17 +47,17 @@ def init_client(cls, host, distance, connection_params: dict, search_params: dic
...
                 "vector": {
-                    "vector": vector,
+                    "vector": query.vector,
                     "k": top,
                 }
             }
         }
 
-        meta_conditions = cls.parser.parse(meta_conditions)
+        meta_conditions = cls.parser.parse(query.meta_conditions)
```
Comment (joein) [inline engine/clients/opensearch/search.py:60]: query here is not `Query`, but the dict, which has just defined above

File: engine/clients/weaviate/search.py [hunk 7]
```diff
@@ -32,10 +32,10 @@ def init_client(cls, host, distance, connection_params: dict, search_params: dic
         cls.client = client
 
     @classmethod
-    def search_one(self, vector, meta_conditions, top) -> List[Tuple[int, float]]:
+    def search_one(self, query: Query, top: int) -> List[Tuple[int, float]]:
```
Comment (joein) [inline engine/clients/weaviate/search.py:35]: classmethod uses `self` instead of `cls`

File: engine/clients/redis/search.py [hunk 8]
```diff
@@ -41,16 +42,16 @@ def init_client(cls, host, distance, connection_params: dict, search_params: dic
         cls._ft = cls.conns[random.randint(0, len(cls.conns)) - 1].ft()
```
Comment (joein) [inline engine/clients/redis/search.py:42]: `cls.conns`, `cls._is_cluster`, `cls._ft`, `cls.knn_conditions` require definitions on class level (values are not required, only type hints)

**Review by joein — CHANGES_REQUESTED** [review]

File: engine/clients/weaviate/search.py (hunk 7 above)
Comment (KShivendu) [reply engine/clients/weaviate/search.py:35]: Good catch.

File: engine/clients/redis/search.py [hunk 9]
```diff
@@ -19,27 +20,35 @@
...
                 cls.client.get_redis_connection(node)
                 for node in cls.client.get_primaries()
             ]
-        cls._ft = cls.conns[random.randint(0, len(cls.conns)) - 1].ft()
+        else:
+            cls.conns = [cls.client]
+
+        cls.is_cluster = REDIS_CLUSTER
+        cls.search_namespace = random.choice(cls.conns).ft()
+        cls.knn_conditions = "EF_RUNTIME $EF"
```
Comment (joein) [inline engine/clients/redis/search.py:51]: btw, if it is a class constant, i think it should be defined on the class level and not in init_client

**Review by joein — APPROVED** [review]

**Pull request merged**